repository: increase test timeout for repository analysis

schase-es opened the pull request:
The MinIO repository analyze test is failing in builds sometimes because of
a timeout. This change increases it from 2 minutes to 5 minutes.

Closes: #134853

schase-es committed:
repository: increase timeout for repository analysis

The MinIO repository analyze test is failing in builds sometimes because of a
timeout. This change increases it from 2 minutes to 5 minutes.

schase-es committed:
Merge branch 'main' into test_failure/134853-minio-repository-analysis

schase-es committed:
Unmuting tests

DaveCTurner reviewed (changes requested):
Hmm I suspect this test failure might represents a continuation of the saga in MinIO around concurrent `UploadPart` and `AbortMultipartUpload` calls:

- https://github.com/minio/minio/issues/21189
- https://github.com/minio/minio/issues/21377
- https://github.com/minio/minio/issues/21456

The change that closed https://github.com/minio/minio/issues/21456 might not actually have fixed the issue, it might have just made it so that retrying the `AbortMultipartUpload` call eventually cleans things up. I'd rather we ruled out such issues in MinIO before just blindly increasing the timeout. 2 minutes should already be ample time to complete this test.

schase-es committed:
Merge branch 'main' into test_failure/134853-minio-repository-analysis

schase-es commented:
@DaveCTurner thanks for the added context -- I'm realizing I need to do some searching around and git-blaming to look at things like this more. I did need to get docker setup before I could test this locally -- so I was very blind.

The docker image for testing this is the 9/07 release. Most of the tests are around 20 seconds. But the reproduction string is running at about 3'20":
```./gradlew ":x-pack:plugin:snapshot-repo-test-kit:qa:minio:javaRestTest" --tests "org.elasticsearch.repositories.blobstore.testkit.analyze.MinioRepositoryAnalysisRestIT.testRepositoryAnalysis" -Dtests.seed=43BF8E5368D66EDC -Dtests.locale=asa-TZ -Dtests.timezone=America/Thunder_Bay -Druntime.java=24
```

I tried:
- adding back in the settings from the PR you linked: time_to_live and anti_contention_delay
- increasing settings like throttled_delete_retry.delay_increment and throttled_delete_retry.delay_increment
- lowering the cooldown_period to one minute (three minutes and 20 would be perfect for hitting this)
- in S3BlobContainer::get_register, temporarily removing the noBackoff policy for the analysis pathway

I'm still seeing 3'20" with various combinations of these.

Do you have any ideas about what this could be hitting?

DaveCTurner commented:
Nothing obvious, sorry, I would either add logging or take packet captures to work out what's unusual about these slower runs.

schase-es commented:
Capturing the packets was a really good idea. It's not visible from the elastic logs, but minio returns a 404 file-not-found to the copy operation, and a 503 unavailable to the put operation. Both responses come after 30 seconds, with the delay on the minio server side.

The 503 causes the s3 client to sleep for a minute before retrying the put, which does succeed. So the delay is partly minio, and partly inside the client.

DaveCTurner commented:
Are you saying the `CopyObject` operation is requested before the `PutObject` completes? If so, a `404` is acceptable, but it should be returned promptly, and it's not expected that a concurrent `CopyObject` can cause a concurrent `PutObject` to fail with a `503`. The client's minute-long pause may be desired behaviour, depending on the exact body of the `503` response - if it's not a valid S3 error response then backing off hard and then retrying seems like a reasonable sort of reaction.

schase-es commented:
Yes, that's right -- the `CopyObject` is started while the `PutObject` is almost done. It's all in one thread though.

The specific sequence is:
- `BlobAnalyzeAction.writeRandomBlob` starts a `PutObject` to key A, and streams most of its data
- when the last byte is requested from the write input stream, it calls into `BlobAnalyzeAction.onLastReadForInitialWrite`, which starts the `CopyObject` from A -> B. The key for A doesn't exist yet, so this can't work
- 30 seconds later, both hear back from the server, with a 404 for the copy and a 503 for the put object
- one minute later, the `PutObject` is retried and succeeds

The repository analyze test does do this copy just before the last byte sometimes -- it's randomized, and the test seed that was reproducible happened to be doing it twice. This caused the test to timeout.

The modified test only does this (plus the repository verification) to make data and logs straightforward.

DaveCTurner commented:
Ok cool. Does this always happen if we try to `CopyObject` during a `PutObject` or is there some race on the MinIO side too?

schase-es commented:
Yes, it does always fail like this -- that was one of the odder initial things about the reproduction seed, was that it happened 100% of the time. It's hard to know though, because the timing is also so consistent from the elastic side.

I'm much less well-versed in locks and consensus, but this seems right: There are two parties interested in the lock: the writer, which is well established when the copy comes in. From reading the responses and timing, their current solution appears to be that the copy is queued up to wait for the lock, which eventually times out the writer and strips the lock. Minio may be able to do an optimization to check the key without the lock first.

DaveCTurner commented:
I don't think we need to get into the depths of the MinIO implementation in terms of locks etc - it should be enough to describe the symptoms we see in terms of the HTTP requests & responses. For the last few issues I've raised with them it's not been possible to reproduce the issue very reliably, but it sounds like we could even create a small test program that demonstrates the issue here. That'd be very helpful to them.

schase-es commented:
Minio is not interested in this delay, and views it as expected given how PutObject holds a lock while running. (https://github.com/minio/minio/issues/21683#issuecomment-3506436504)

I can increase the test timeout for minio to 4 minutes, or remove this part of the test for minio.

DaveCTurner commented:
I think maybe they have missed the point that this delay is an egregious
30s and the PutObject ultimately fails. If it were just waiting for the
last byte and the request eventually succeeded then it'd be different.

On Mon, 10 Nov 2025, 01:20 Simon Chase, ***@***.***> wrote:

> *schase-es* left a comment (elastic/elasticsearch#135565)
> <https://github.com/elastic/elasticsearch/pull/135565#issuecomment-3509077296>
>
> Minio is not interested in this delay, and views it as expected given how
> PutObject holds a lock while running. (minio/minio#21683 (comment)
> <https://github.com/minio/minio/issues/21683#issuecomment-3506436504>)
>
> I can increase the test timeout for minio to 4 minutes, or remove this
> part of the test for minio.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/135565#issuecomment-3509077296>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABGS53FGUNVFQQ452EHHHLL337R7NAVCNFSM6AAAAACHUTYVGWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMBZGA3TOMRZGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

DaveCTurner commented:
> PutObject will hold the write lock while it is running

... or the point that AWS S3 doesn't have any such lock. It might be expected based on their implementation but it's a deviation from S3's behaviour so they can't really claim to be S3-compatible if they take this line.

schase-es committed:
repository: turn off copy-during-write for minio

Minio's implementation of PutObject means BlobAnalyzeAction's early copy during
write has a long wait, followed by a timeout. Together, this is 90 seconds,
which can time out the test whenever it happens twice. This change adds a
setting to turn off the copy during write for Minio.

schase-es committed:
Merge branch 'main' into test_failure/134853-minio-repository-analysis

schase-es committed:
Un-increase timeout for repository test

Commit:
[CI] Auto commit changes from spotless

schase-es committed:
Remove extra tests added in botched github merge

Note to self: never resolve conflicts in github

DaveCTurner reviewed (commented):
Makes sense, thanks for investigating here.

schase-es committed:
Change source from ClusterSettings to Settings

schase-es committed:
Merge branch 'main' into test_failure/134853-minio-repository-analysis

DaveCTurner commented on x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/analyze/BlobAnalyzeAction.java:
```diff
@@ -171,7 +173,19 @@
+        Setting.Property.NodeScope,
+        Setting.Property.Deprecated
+    );
+    private volatile boolean enableEarlyCopy = true;
```
```suggestion
    private final boolean enableEarlyCopy = true;
```

DaveCTurner reviewed (approved):
LGTM (one nit)

schase-es committed:
Adding nit

schase-es committed:
remove init

schase-es committed:
Merge branch 'main' into test_failure/134853-minio-repository-analysis
